AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/handler/BeanHandler.java

```diff
@@ -25,52 +22,58 @@
 import org.androidannotations.ElementValidation;
 import org.androidannotations.annotations.Bean;
 import org.androidannotations.annotations.EBean;
-import org.androidannotations.annotations.NonConfigurationInstance;
 import org.androidannotations.handler.BaseAnnotationHandler;
+import org.androidannotations.handler.MethodInjectionHandler;
+import org.androidannotations.helper.InjectHelper;
 import org.androidannotations.holder.EBeanHolder;
 import org.androidannotations.holder.EComponentHolder;
 
 import com.helger.jcodemodel.AbstractJClass;
+import com.helger.jcodemodel.IJExpression;
 import com.helger.jcodemodel.JBlock;
-import com.helger.jcodemodel.JFieldRef;
-import com.helger.jcodemodel.JInvocation;
 
-public class BeanHandler extends BaseAnnotationHandler<EComponentHolder> {
+public class BeanHandler extends BaseAnnotationHandler<EComponentHolder>implements MethodInjectionHandler<EComponentHolder> {
+
+	private final InjectHelper<EComponentHolder> injectHelper;
 
 	public BeanHandler(AndroidAnnotationsEnvironment environment) {
 		super(Bean.class, environment);
+		injectHelper = new InjectHelper<>(validatorHelper, this);
 	}
 
 	@Override
 	public void validate(Element element, ElementValidation validation) {
-		validatorHelper.enclosingElementHasEnhancedComponentAnnotation(element, validation);
+		injectHelper.validate(Bean.class, element, validation);
+
+		validatorHelper.typeOrTargetValueHasAnnotation(EBean.class, element, validation);
 
 		validatorHelper.isNotPrivate(element, validation);
+	}
 
-		validatorHelper.typeOrTargetValueHasAnnotation(EBean.class, element, validation);
+	@Override
+	public void process(Element element, EComponentHolder holder) {
+		injectHelper.process(element, holder);
+	}
+
+	@Override
+	public JBlock getInvocationBlock(EComponentHolder holder) {
+		return holder.getInitBodyInjectionBlock();
 	}
 
 	@Override
-	public void process(Element element, EComponentHolder holder) throws Exception {
+	public IJExpression getInstanceInvocation(Element element, EComponentHolder holder, Element param) {
 		TypeMirror typeMirror = annotationHelper.extractAnnotationClassParameter(element);
 		if (typeMirror == null) {
-			typeMirror = element.asType();
+			typeMirror = param.asType();
 			typeMirror = getProcessingEnvironment().getTypeUtils().erasure(typeMirror);
 		}
-
 		String typeQualifiedName = typeMirror.toString();
 		AbstractJClass injectedClass = getJClass(annotationHelper.generatedClassQualifiedNameFromQualifiedName(typeQualifiedName));
+		return injectedClass.staticInvoke(EBeanHolder.GET_INSTANCE_METHOD_NAME).arg(holder.getContextRef());
+	}
 
-		String fieldName = element.getSimpleName().toString();
-		JFieldRef beanField = ref(fieldName);
-		JBlock block = holder.getInitBodyInjectionBlock();
-
-		boolean hasNonConfigurationInstanceAnnotation = element.getAnnotation(NonConfigurationInstance.class) != null;
-		if (hasNonConfigurationInstanceAnnotation) {
```

Comment:
hm, i thought i moved it somewhere else. but it looks like it got lost... need to find a way to support an edge case like this... (similar to the logging :x)